### PR TITLE
Api refactoring

### PR DESCRIPTION
**Pull request by inakianduaga**
# Description

- Document client config options
- Simplify extractors
- Add library for generating request/responses stubs without hacks during tests
- Extractors support multiparam syntax together with pipe syntax
`?toguru=foo=true&toguru=bar=false` or `?toguru=foo=true|bar=false`. Same for header values

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -5,46 +5,75 @@ import { UserInfo, ToguruData } from './models/toguru'
...
-    togglesForService: (service: string, user: UserInfo) => Toggles
+    /**
+     * Determine if the toggle is enabled based on user information
+     */
+    isToggleEnabled(user: UserInfo): (toggle: Toggle) => boolean
+    /**
+     * Determine if the toggle is enabled for a given request and optional extractor.
+     * If extractor is not provided, it will default to standard one
+     */
+    isToggleEnabled(request: Request, extractor: (request: Request) => UserInfo): (toggle: Toggle) => boolean
```
Comment (mriceron) [inline src/client.ts]: I'm not sure if I would put some integration code with a framework on the client.

We could have a `ClientWithInfo` or whatever that take care of extracting the `UserInfo` and expose only a `isToggleEnabled(toggle: Toggle) => boolean` kind of like the middleware does.

And we could reuse the code in the middleware itself

Comment (inakianduaga) [reply src/client.ts]: I reverted the client change, will add something on the express middleware.

File: src/expressMiddleware/extractors.ts [hunk 2]
```diff
@@ -1,57 +1,72 @@
...
+    const value = rawCookie ? valueCaseInsensitiveKey(name, cookie.parse(rawCookie)) : null
+
+    return value ? deserializeToguruValue(value) : {}
+}
 
-    const togglesFromQueryParam = () => {
-        const toguruQuery = valueCaseInsensitiveKey('toguru', request.query ?? {})
-        return toguruQuery ? togglesStringParser(toguruQuery) : undefined
-    }
+export const fromHeader = (name: string): ForcedToggleExtractor => (request) => {
```
Comment (mriceron) [inline src/expressMiddleware/extractors.ts]: If we move all the extractors to `ForcedToggleExtractor` they won't be usable for attributes and the uuid.

File: src/expressMiddleware/extractors.ts [hunk 3]
```diff
@@ -1,57 +1,72 @@
...
+        ? value.map((v) => deserializeToguruValue(v)).reduce((acc, c) => ({ ...acc, ...c }))
+        : value
+        ? deserializeToguruValue(value)
+        : {}
+}
+
+/**
+ * Fetches forced toggles from a request based on defaults
+ */
+export const defaultForcedTogglesExtractor = (request: Request): Record<string, boolean> => ({
```
Comment (mriceron) [inline src/expressMiddleware/extractors.ts]: Here we would differ with the Scala implementation where there can be only a single source for forced toggles.
The order in which they are evaluated is => `QueryParam -> Header -> Cookie` (that was incorrect before)

Comment (inakianduaga) [reply src/expressMiddleware/extractors.ts]: > The order in which they are evaluated is => QueryParam -> Header -> Cookie (that was incorrect before)

Will change the order

> Here we would differ with the Scala implementation where there can be only a single source for forced toggles.

I don't see why this should be a requirement that the value is specified just in one place. If you have a cookie value for another toggle and you override something via queryParam then you get the combination. I talked to @rtimush quickly and he agrees.

Comment (mriceron) [reply src/expressMiddleware/extractors.ts]: > I don't see why this should be a requirement that the value is specified just in one place. If you have a cookie value for another toggle and you override something via queryParam then you get the combination. I talked to @rtimush quickly and he agrees.
It's totally fine for me 👍  
I was just pointing out that we will differ from the Scala implementation

File: src/expressMiddleware/extractors.ts (hunk 2 above)
Comment (inakianduaga) [reply src/expressMiddleware/extractors.ts]: there is a small separate one that reads a value from a cookie, that's the only one that's used for the `culture` and the `uuid`.

**Pull request merged**